json/src/main/java/io/airlift/json/JsonCodec.java

```diff
@@ -214,6 +218,9 @@ public byte[] toJsonBytes(T instance)
             throws IllegalArgumentException
     {
         try {
+            if (isJsonSubType(mapper, instance)) {
+                return mapper.writeValueAsBytes(instance);
+            }
```

Comment:
i'm generally ok with this if it's how it has to be, but it doesn't _feel_ like it should be necessary, as `fromJson` always uses a reader, and it works just fine when reading the subtype (probably because we `addDeserializer` to the module). This just feels like it shouldn't be necessary - like there should be a way to implement a serializer that Just Works, though I could not find one.

I'm also totally cool with removing the readers and writers from the equation altogether, but that feels risky regarding downstream effects.